src/blocked.c

```diff
@@ -608,6 +607,11 @@ void handleClientsBlockedOnKeys(void) {
                  * regardless of the object type: we don't know what the
                  * module is trying to accomplish right now. */
                 serveClientsBlockedOnKeyByModule(rl);
+            } else {
+                /* Edge case: If lookupKeyReadWithFlags decides to expire the key we have to
+                 * take care of the propagation here, because afterCommand wasn't called */
+                if (server.also_propagate.numops > 0)
+                    propagatePendingCommands();
```

Comment:
good catch!!!
i see we already had a test for that? so how come we didn't get the assertion before?

p.s. another way to test it could be:
MULTI + LPUSH + EXPIRE + DEBUG SLEEP

Reply:
3 commits ago i added the assertion `serverAssert(server.also_propagate.numops == 0);` in `handleClientsBlockedOnKeys` and then the external-server tests failed on that SWAPDB test

the reason it failed is that `SWAPDB` called `handleClientsBlockedOnKeys` which added the expire-delete to also_propagate (and never actually propagated) and then `handleClientsBlockedOnKeys` was called again from `beforeSleep`, triggering the assertion.

the reason It happened only with the external server is probably that at some point the external server loaded an RDB, causing it to create a backlog and thus `shouldPropagate` returned 1 (in non-external tests the server come up fresh, and if there's no AOF or replicas, `alsoPropagate` doesn't do anything)

Reply:
so maybe it's wise to use this opportunity to add an explicit test for it (MULTI + LPUSH + EXPIRE + DEBUG SLEEP)?

Reply:
i actually edited to original test to check the replication stream, so now even if it's not in standalone mode, the server still replicates... i checked and indeed it failed without the fix in non-standalone mode

do you still want me to add the MULTI + LPUSH + EXPIRE + DEBUG SLEEP case?

Reply:
this area of redis seems fragile, so i think more tests to detect future regressions are better.
but if it takes too long to compose that test, i'm willing to let go.

Reply:
i'll add the test, should be easy enough